openfe_gromacs/protocols/gromacs_md/md_methods.py

```diff
@@ -0,0 +1,649 @@
+        repeat_id : int
+          identifier for which repeat (aka replica/clone) this Unit is
+        generation : int
+          counter for how many times this repeat has been extended
+        name : str, optional
+          human-readable identifier for this Unit
+
+        Notes
+        -----
+        The mapping used must not involve any elemental changes.  A check for
+        this is done on class creation.
+        """
+        super().__init__(
+            name=name,
+            protocol=protocol,
+            stateA=stateA,
+            repeat_id=repeat_id,
+            generation=generation,
+        )
+
+    @staticmethod
+    def _assign_partial_charges(
+        charge_settings: OpenFFPartialChargeSettings,
+        smc_components: dict[SmallMoleculeComponent, OFFMolecule],
+    ) -> None:
+        """
+        Assign partial charges to SMCs.
+
+        Parameters
+        ----------
+        charge_settings : OpenFFPartialChargeSettings
+          Settings for controlling how the partial charges are assigned.
+        smc_components : dict[SmallMoleculeComponent, openff.toolkit.Molecule]
+          Dictionary of OpenFF Molecules to add, keyed by
+          SmallMoleculeComponent.
+        """
+        for mol in smc_components.values():
+            charge_generation.assign_offmol_partial_charges(
+                offmol=mol,
+                overwrite=False,
+                method=charge_settings.partial_charge_method,
+                toolkit_backend=charge_settings.off_toolkit_backend,
+                generate_n_conformers=charge_settings.number_of_conformers,
+                nagl_model=charge_settings.nagl_model,
+            )
+
+    def run(
+        self, *, dry=False, verbose=True, scratch_basepath=None, shared_basepath=None
+    ) -> dict[str, Any]:
+        """Run the MD simulation.
+
+        Parameters
+        ----------
+        dry : bool
+          Do a dry run of the calculation, creating all necessary hybrid
+          system components (topology, system, sampler, etc...) but without
+          running the simulation.
+        verbose : bool
+          Verbose output of the simulation progress. Output is provided via
+          INFO level logging.
+        scratch_basepath: Pathlike, optional
+          Where to store temporary files, defaults to current working directory
+        shared_basepath : Pathlike, optional
+          Where to run the calculation, defaults to current working directory
+
+        Returns
+        -------
+        dict
+          Outputs created in the basepath directory or the debug objects
+          (i.e. sampler) if ``dry==True``.
+
+        Raises
+        ------
+        error
+          Exception if anything failed
+        """
+        if verbose:
+            self.logger.info("Creating system")
+        if shared_basepath is None:
+            # use cwd
+            shared_basepath = pathlib.Path(".")
+
+        # 0. General setup and settings dependency resolution step
```

Comment:
I would probably split this into three methods call to keep it clean.

Reply:
I split these up into separate methods.